service/history/workflow/state_machine_timers.go

```diff
@@ -55,16 +56,23 @@ func AddNextStateMachineTimerTask(ms MutableState) {
 
 // TrackStateMachineTimer tracks a timer task in the mutable state's StateMachineTimers slice sorted and grouped by
 // deadline.
+// Tasks are deduplicated using proto equality.
 func TrackStateMachineTimer(ms MutableState, deadline time.Time, taskInfo *persistencespb.StateMachineTaskInfo) {
 	execInfo := ms.GetExecutionInfo()
 	group := &persistencespb.StateMachineTimerGroup{
 		Deadline: timestamppb.New(deadline),
 		Infos:    []*persistencespb.StateMachineTaskInfo{taskInfo},
 	}
-	idx, found := slices.BinarySearchFunc(execInfo.StateMachineTimers, group, func(a, b *persistencespb.StateMachineTimerGroup) int {
+	idx, groupFound := slices.BinarySearchFunc(execInfo.StateMachineTimers, group, func(a, b *persistencespb.StateMachineTimerGroup) int {
 		return a.Deadline.AsTime().Compare(b.Deadline.AsTime())
 	})
-	if found {
+	if groupFound {
+		taskFound := slices.ContainsFunc(execInfo.StateMachineTimers[idx].Infos, func(info *persistencespb.StateMachineTaskInfo) bool {
+			return proto.Equal(info, taskInfo)
```

Comment:
hmmm I think we will need to exclude some fields like lastUpdateVersionedTransition or MachineTransitionCount or the two taskInfos won't be equal?

But then task may not be actual duplication? maybe we can always pick the one with higher machineTransitionCount, i.e. replace the taskInfo

Reply:
Fixed.